cadc-storage-adapter/src/main/java/org/opencadc/inventory/storage/StorageClient.java

```diff
@@ -0,0 +1,195 @@
+ *
+ *  $Revision: 4 $
+ *
+ ************************************************************************
+ */
+
+package org.opencadc.inventory.storage;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.StreamCorruptedException;
+import java.lang.reflect.Constructor;
+import java.net.URI;
+import java.util.Iterator;
+
+import org.apache.log4j.Logger;
+import org.opencadc.inventory.Artifact;
+import org.opencadc.inventory.StorageLocation;
+
+import ca.nrc.cadc.net.InputStreamWrapper;
+import ca.nrc.cadc.net.OutputStreamWrapper;
+import ca.nrc.cadc.net.ResourceNotFoundException;
+import ca.nrc.cadc.net.TransientException;
+
+/**
+ * Provides access to storage.  
+ *
+ * @author majorb
+ *
+ */
+public class StorageClient {
+    
+    private static Logger log = Logger.getLogger(StorageClient.class);
+    
+    private StorageAdapter adapter;
+
+    public StorageClient() {
+        adapter = getStorageAdapter();
+    }
+    
+    public Artifact get(URI storageID, OutputStream out) throws ResourceNotFoundException, TransientException {
+        InputStreamWrapper handler = new InputStreamWrapper() {
+            public void read(InputStream in) throws IOException {
+                ioLoop(out, in);
+            }
+        };
+        return adapter.get(storageID, handler);
+    }
+
+    public Artifact meta(URI storageID) throws ResourceNotFoundException, TransientException {
+        return adapter.meta(storageID);
+    }
+
+    public StorageLocation put(Artifact artifact, InputStream in) throws StreamCorruptedException, TransientException {
+        OutputStreamWrapper wrapper = new OutputStreamWrapper() {
+            public void write(OutputStream out) throws IOException {
+                ioLoop(out, in);
+            }
+        };
+        return adapter.put(artifact, wrapper);
+    }
+
+    public StorageLocation replace(Artifact artifact, InputStream in, StorageLocation replaceID) throws ResourceNotFoundException, StreamCorruptedException, TransientException {
+        OutputStreamWrapper wrapper = new OutputStreamWrapper() {
+            public void write(OutputStream out) throws IOException {
+                ioLoop(out, in);
+            }
+        };
+        return adapter.replace(artifact, wrapper, replaceID);
+    }
+
+    public void delete(URI storageID) throws ResourceNotFoundException, TransientException {
+        adapter.delete(storageID);
+    }
+
+    public Iterator<StorageLocation> iterator() throws TransientException {
+        return adapter.iterator();
+    }
+    
+    private void ioLoop(OutputStream out, InputStream in) {
+        // TODO: Write 2 threaded io loop--one thread reading
+        // and one thread writing.
+    }
+    
+    /**
+     * Load the configured storage adapter or the default one if none
+     * are found.
+     * 
+     * @return The storage adapter to be used by this client.
+     */
+    private StorageAdapter getStorageAdapter() {
```

Comment:
These will be configurable by system property only (using catalina.properties) so that we can make a fully reusable docker image. We may want to figure out how to achieve that without having to include all storage-adapter implementation jars (which could get pretty fat) but that's not a concern for now.

Change javadoc to remove "default".